### ISSUE TITLE
Proposal: DependsOnTask for project items Include

### ISSUE THREAD

**Issue by kasper3** [Microsoft/msbuild#3220]
For `Problem statement` and `Part 1`, see https://github.com/Microsoft/msbuild/issues/3219

Part 2:

```xml
<Compile
  Include="$(MSBuildProjectDirectory)\SharedDTOs\RemoteFile.vb"
  DependsOnTask="DownloadMySourceFile"
  Condition="!Exist($(MSBuildProjectDirectory)\SharedDTOs\RemoteFile.vb)" />
```

this will indicate the design-time or first-evaluation-time to run the task before evaluating this line, ignore the task result and continue evaluation once task is finished. If task fails (due to insufficient credentials / no internet connection / wrong URL / any other reason), a reason warning will be issued and subsequent steps will fail as they fail today:

> CSC : error CS2001: Source file 'C:\testproj\nonexisting.cs' could not be found.

for `<Compile>` and for `<Content>`

> error : File not found: 'C:\testproj\mockimage.png'

etc.

**Comment by rainersigwald** [comment]
I'm not so sure about this. Would it apply to evaluation-time (as opposed to created-during-execution) items? That is, is it legal to have

```xml
<Project>
  <ItemGroup>
    <I Include="foo.bar" DependsOnTask="CreateFoo" />
  </ItemGroup>
</Project>
```

? What about

```xml
<Project InitialTargets="IncludeFoo">
  <Target Name="IncludeFoo">
    <ItemGroup>
      <I Include="foo.bar" DependsOnTask="CreateFoo" />
    </ItemGroup>
  </Target>
</Project>
```

? How does the execution differ?

Do you want to call a task or a target with this? If a task, how would you pass arguments to the task, or would it be a requirement that the task not take any arguments?

**Comment by CZEMacLeod** [comment]
I think you might need some kind of metabase to store the etag, last modified, cache control settings of the downloaded file. Also control over when the file is updated - some kind of max query rate control that can be applied by the consumer. If this is just to restore missing files, that is one thing, but if it is to download and keep a file up to date that is another...
Is it applicable to have DownloadFiles as an ItemGroup and maybe make it visible separately in the UI (e.g. Visual Studio) with a property page to set the metadata.
I think the downloaded files would be stored in the IntermediateOutputFiles path or similar...

```
<ItemGroup>
	<DownloadFile Include="https://my.company.com/logo.png">
		<TargetGroup>Content</TargetGroup>
		<MaxUpdate>86400</MaxUpdate>
		<Pack>true</Pack>
		<PackagePath>content\images\</PackagePath>
	</DownloadFile>
</ItemGroup>
```

If using that approach, then I think you could make the `DownloadFiles` target just be `BeforeTargets="Build"`?

**Comment by CZEMacLeod** [comment]
@kasper3 The Pack, and PackagePath properties were there as an example to indicate that all metadata was copied to the appropriate TargetGroup ItemGroup in the DownloadFiles target. 
I was thinking something like a resolve downloads, where the filename would be inferred from the url unless it was specified in via content-disposition etc.
This would probably act a little like the Microsoft.NuGet.targets ResolveNuGetPackageAssets target.

In the project UI, deferred, or remote items should be able to be downloaded, refreshed or force refreshed.
